File: tools/http/test-client/java/src/main/java/io/opentelemetry/conformance/http/HttpClientWorkload.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */
package io.opentelemetry.conformance.http;

import io.opentelemetry.conformance.http.HttpContract.Exchange;
import io.opentelemetry.conformance.http.HttpContract.Response;
import java.time.Duration;

/**
 * Shared support for JVM client scenarios: the request contract, sent by the library under test.
 *
 * <p>Only a <em>client</em> scenario needs this: it is the sender, so the requests have to leave
 * the library under test. A server scenario is driven from outside its own process by {@code
 * otel-http-drive} and never sends anything.
 *
 * <p>The shared telemetry contract checks what this request emits. The helper also verifies the
 * response against the selected contract entry.
 */
public final class HttpClientWorkload {

  /** Maximum time a client scenario waits for one request to finish. */
  public static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(30);

  private static final Duration INSTRUMENTATION_QUIESCENCE = Duration.ofMillis(100);

  private HttpClientWorkload() {}

  /** Sends one request using the HTTP client library under test. */
  @FunctionalInterface
  public interface Sender {
    /** {@code body} is null for a request that carries none. */
    Response send(String method, String url, String body) throws Exception;
  }

  /**
   * Sends the runner-selected contract request at {@code baseUrl} through {@code sender}.
   *
   * <p>No health check: the runner starts the mock server a client scenario calls and waits for it
   * to answer before running the scenario at all.
   */
  public static void drive(String baseUrl, Sender sender) throws Exception {
    drive(baseUrl, sender, HttpContract.scenarioRequest());
  }

  static void drive(String baseUrl, Sender sender, Exchange exchange) throws Exception {
    if (baseUrl.isBlank()) {
      throw new IllegalArgumentException("base URL must not be blank");
    }
    Response response = sender.send(exchange.method(), baseUrl + exchange.path(), exchange.body());
    System.out.printf(
        "%s %s -> %d %s%n",
        exchange.method(), exchange.path(), response.statusCode(), abbreviate(response.body()));
    HttpContract.verify(exchange, response);
    // Some asynchronous clients return the body before their instrumentation completion callback.
    Thread.sleep(INSTRUMENTATION_QUIESCENCE.toMillis());
  }

  private static String abbreviate(String value) {
    String singleLine = value.replace('\r', ' ').replace('\n', ' ');
    return singleLine.length() <= 60 ? singleLine : singleLine.substring(0, 60);
  }
}

File: tools/http/test-client/dotnet/OpenTelemetry.Conformance.Http/HttpClientWorkload.cs
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

namespace OpenTelemetry.Conformance.Http;

/// <summary>
/// Shared support for .NET client scenarios: the request contract, sent by the library under test.
/// </summary>
/// <remarks>
/// Only a <em>client</em> scenario needs this: it is the sender, so the requests have to leave the
/// library under test. A server scenario is driven from outside its own process by
/// <c>otel-http-drive</c> and never sends anything.
/// <para>The shared telemetry contract checks what these requests emit.</para>
/// </remarks>
public static class HttpClientWorkload
{
    /// <summary>Sends one request using the HTTP client library under test.</summary>
    /// <remarks><paramref name="body"/> is null for a request that carries none.</remarks>
    public delegate Task<HttpContract.Response> Sender(string method, string url, string? body);

    /// <summary>
    /// Sends the runner-selected contract request at <paramref name="baseUrl"/> through
    /// <paramref name="send"/>.
    /// </summary>
    /// <remarks>
    /// No health check: the runner starts the mock server a client scenario calls and waits for it
    /// to answer before running the scenario at all.
    /// </remarks>
    public static async Task DriveAsync(string baseUrl, Sender send)
    {
        await DriveAsync(baseUrl, send, HttpContract.ScenarioRequest()).ConfigureAwait(false);
    }

    internal static async Task DriveAsync(
        string baseUrl,
        Sender send,
        HttpContract.Exchange exchange)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(baseUrl);
        ArgumentNullException.ThrowIfNull(send);

        var normalizedBaseUrl = baseUrl.TrimEnd('/');
        var response = await send(
                exchange.Method, normalizedBaseUrl + exchange.Path, exchange.Body)
            .ConfigureAwait(false);
        Console.WriteLine(
            $"{exchange.Method} {exchange.Path} -> {response.StatusCode} {Abbreviate(response.Body)}");
    }

    private static string Abbreviate(string value)
    {
        var singleLine = value.ReplaceLineEndings(" ");
        return singleLine.Length <= 60 ? singleLine : singleLine[..60];
    }
}
